namerd/iface/destination/src/main/scala/io/buoyant/namerd/iface/destination/DestinationService.scala

```diff
@@ -0,0 +1,150 @@
+package io.buoyant.namerd.iface.destination
+
+import com.twitter.finagle.naming.NameInterpreter
+import com.twitter.finagle.{Address, _}
+import com.twitter.io.Buf
+import com.twitter.logging.Logger
+import com.twitter.util._
+import io.buoyant.grpc.runtime.{GrpcStatus, Stream}
+import io.linkerd.proxy.destination.Destination
+import io.linkerd.proxy.destination.Update.OneofUpdate
+import io.linkerd.proxy.destination.Update.OneofUpdate.{NoEndpoints => EndpointsNone, _}
+import io.linkerd.proxy.destination._
+import io.linkerd.proxy.net.IPAddress.OneofIp
+import io.linkerd.proxy.net.{IPAddress, TcpAddress}
+import java.net.{InetAddress, InetSocketAddress}
+
+class DestinationService(
+  interpreter: NameInterpreter
+) extends Destination {
+  import DestinationService._
+
+  //Diff implementation for Address Set
+  class AddressDiffable extends Diffable[Set] {
+    override def diff[T](
+      left: Set[T],
+      right: Set[T]
+    ): Diff[Set, T] = AddressDiff(right -- left, left -- right)
+
+    override def empty[T]: Set[T] = Set.empty
+  }
+
+  case class AddressDiff[T](add: Set[T], remove: Set[T]) extends Diff[Set, T] {
+    override def patch(coll: Set[T]): Set[T] = coll ++ add -- remove
+
+    override def map[U](f: T => U): Diff[Set, U] = AddressDiff(add.map(f), remove.map(f))
+  }
+
+  implicit val _addressDiffable = new AddressDiffable
+
+  private[this] val log = Logger.get(getClass.getName)
+  override def get(req: GetDestination): Stream[Update] = {
+    log.info(s"stream initiated from request: $req")
+    if (!req.scheme.contains("k8s")) {
+      val res = Stream.mk[Update]
+      res.close(GrpcStatus.Unimplemented("Unknown scheme: " + req.scheme.getOrElse("unknown")))
+      return res
+    }
+
+    // This assumes that our dtab begins with the /svc prefix
+    val name = Path(Seq(Buf.Utf8("svc")) ++ req.path.map(Buf.Utf8(_)): _*)
+
+    val stream = Stream.mk[Update]
+    interpreter.bind(Dtab.empty, name).run.changes.respond {
+      case Activity.Pending => // Immediately send no endpoints when stream starts successfully
+        stream.send(mkNoEndpointsUpdate(false)); ()
+      case Activity.Failed(e) => // Send no endpoints message when we fail to get endpoints
+        log.info(s"address lookup failed: $e")
+        stream.send(mkNoEndpointsUpdate(false)); ()
+      case Activity.Ok(t) => t.eval match {
+        case None => // Service exists but cannot find addresses for service.
+          stream.send(mkNoEndpointsUpdate(true)); ()
+        case Some(value) => // Evaluate what diff of addresses and send add, remove and update requests.
+          val _ = updates(name, value).diff.respond {
+            case AddressDiff(add, remove) =>
+              if (add.nonEmpty) stream.send(mkAddUpdate(add)); ()
```

Comment:
This is a wrinkle when trying to convert Event[AddressDiff] to Event[Update] because one AddressDiff can turn into potentially two Updates.  There's no built-in Event combinator that helps us with this, but we can write our own:

```
def eventFlatMap[T, U](ev: Event[T], f: T => Seq[U]): Event[U] = new Event[U] {
  def register(w: Witness[U]): Closable = {
    ev.respond { t => 
      for(u <- f(t)) w.notify(u)
    }
  }
}
```

and then we can do:

```
val diffs: Event[AddressDiff] = ...
val updates = eventFlatMap(diffs, { case AddressDiff(add, remove) => 
  val addUpdate = if (add.nonEmpty) Seq(mkAddUpdate(add)) else Nil
  val removeUpdate = if (remove.nonEmpty) Seq(mkRemoveUpdate(remove)) else Nil
  addUpdate ++ removeUpdate
}
```